distributed/worker.py

```diff
@@ -3448,14 +3448,14 @@ def _update_who_has(self, who_has: Mapping[str, Collection[str]]) -> None:
                 continue
 
             for worker in ts.who_has - workers:
-                self.has_what[worker].discard(key)
-                # Can't remove from self.data_needed_per_worker; there is logic
-                # in _select_keys_for_gather to deal with this
+                self.has_what[worker].remove(key)
+                if ts.state == "fetch":
+                    self.data_needed_per_worker[worker].remove(ts)
```

Comment:
```suggestion
                self.data_needed_per_worker[worker].discard(ts)
```

Reply:
I'd much rather fail loudly if for any reason the assumption that fetch state and existence in data_needed are inextricably bound together fails